src/Compilers/CSharp/Portable/Syntax/SyntaxNodeExtensions.cs

```diff
@@ -71,7 +71,11 @@ internal static bool CanHaveAssociatedLocalBinder(this CSharpSyntaxNode syntax)
                 kind == SyntaxKind.Attribute ||
                 kind == SyntaxKind.ArgumentList ||
                 kind == SyntaxKind.ArrowExpressionClause ||
-                (syntax is ExpressionSyntax && (syntax.Parent as LambdaExpressionSyntax)?.Body == syntax);
+                (syntax is ExpressionSyntax && 
+                    ((syntax.Parent as LambdaExpressionSyntax)?.Body == syntax ||
+                     (syntax.Parent as SwitchStatementSyntax)?.Expression == syntax ||
+                     (syntax.Parent as ForEachStatementSyntax)?.Expression == syntax ||
+                     (syntax.Parent as IfStatementSyntax)?.Condition == syntax));
```

Comment:
I have a feeling we could benefit from a hypothetical `IsScopeDesignator(syntax)` here.